domaci_bandit.py

```diff
@@ -75,93 +63,107 @@ def take_action(self, a: int):
             return self.bandits[a].pull_leaver()
 
 
-test_env_size = 5
-test_bandits = [Bandit(i**2, i) for i in range(test_env_size)]
-test_env = BanditsEnvironment(test_bandits)
+def choose_greedy_action(q):
+    return np.argmax(q)
 
+def choose_random_action(n):
+    return random.randint(0, n - 1)
 
-selected_bandit = 3
-test_rewards = [test_env.take_action(selected_bandit) for _ in range(test_len)]
 
-plt.plot(test_rewards, label="rewards")
-plt.plot(
-    (selected_bandit**2 + selected_bandit) * np.ones(test_len),
-    linestyle="--",
-    color="r",
-)
-plt.plot(
-    (selected_bandit**2 - selected_bandit) * np.ones(test_len),
-    linestyle="--",
-    color="r",
-)
+def choose_eps_greedy_action(q, eps):
+    if random.random() > eps:
+        return choose_greedy_action(q)
+    else:
+        return choose_random_action(len(q))
 
-selected_bandit = -1
-test_rewards = [test_env.take_action(selected_bandit) for _ in range(test_len)]
 
-plt.plot(test_rewards, label="rewards")
+def train(bandits_no=5, attempts_no=5000, alpha=0.1, epsilon=0.1, plotting=True):
+    bandits = [
+        Bandit(10 * (random.random() - 0.5), 5 * random.random())
+        for _ in range(bandits_no)
+    ]
+    env = BanditsEnvironment(bandits)
 
+    q = [100 for _ in range(bandits_no)]
+    rewards = []
 
-test_rewards = [test_env.take_action(random.randint(0, 4)) for _ in range(test_len)]
-test_mean = sum(test_rewards) / test_len
+    for t in trange(attempts_no):
+        a = choose_eps_greedy_action(q, epsilon)
+        r = env.take_action(a)
+        q[a] = q[a] + alpha * (r - q[a])
 
-print("TEST MEAN = ", test_mean)
+        rewards.append(r)
 
+    if plotting:
+        plt.scatter(range(len(q)), q, marker=".")
+        plt.scatter(range(len(q)), [b.mean for b in env.bandits], marker="x")
 
-def choose_greedy_action(q):
-    return np.argmax(q)
+    return env.bandits, rewards, q
 
 
-test_q = [1, 2, 3, 2, 1]
+# Zadatak 1 : razlicite vrednosti epsilon, prikaz + zakljucak
+def prvi_zadatak():
+    plt.figure(figsize=(14, 3))
 
-test_actions = [choose_greedy_action(test_q) for _ in range(test_len)]
+    plt.subplot(1, 3, 1)
+    envi, rew, q1 = train(epsilon=0.1, plotting=False)
+    g = np.cumsum(rew)
+    max_r = max([b.mean for b in envi])
+    plt.plot(g, "r")
+    plt.plot(np.cumsum(max_r * np.ones(len(g))), "b")
+    plt.title("Epsilon = 0.1")
 
-plt.plot(test_actions)
+    plt.subplot(1, 3, 2)
+    envi, rew, q1 = train(epsilon=0.01, plotting=False)
+    g = np.cumsum(rew)
+    max_r = max([b.mean for b in envi])
+    plt.plot(g, "r")
+    plt.plot(np.cumsum(max_r * np.ones(len(g))), "b")
+    plt.title("Epsilon = 0.01")
 
+    plt.subplot(1, 3, 3)
+    envi, rew, q1 = train(epsilon=0.001, plotting=False)
+    g = np.cumsum(rew)
+    max_r = max([b.mean for b in envi])
+    plt.plot(g, "r")
+    plt.plot(np.cumsum(max_r * np.ones(len(g))), "b")
+    plt.title("Epsilon = 0.001")
+    plt.show()

```

Comment:
Mozes staviti ovo sve u jednu petlju, identican je kod samo menjas subplot index i epsilon, npr. ovako nesto:
`for i in range(3):
    plt.subplot(1, 3, i+1)
    envi, rew, q1 = train(epsilon=1/10^(i+1), plotting=False)
    g = np.cumsum(rew)
    max_r = max([b.mean for b in envi])
    plt.plot(g, "r")
    plt.plot(np.cumsum(max_r * np.ones(len(g))), "b")
    plt.title(f"Epsilon = {1/10^(i+1)}")
    plt.show()`

Reply:
to sam i ja razmisljala da uradim jer se isto to plotovanje koristi i posle u 3 zadatku :)